trin-core/src/types/header.rs

```diff
@@ -244,7 +244,30 @@ pub struct HeaderWithProof {
     pub proof: BlockHeaderProof,
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Decode)]
+impl ssz::Encode for HeaderWithProof {
+    fn is_ssz_fixed_len() -> bool {
+        false
+    }
+
+    fn ssz_append(&self, buf: &mut Vec<u8>) {
+        let header = rlp::encode(&self.header).to_vec();
+        let header: ByteList = ByteList::from(header);
+        let offset =
+            <ByteList as Encode>::ssz_fixed_len() + <AccumulatorProof as Encode>::ssz_fixed_len();
+        let mut encoder = SszEncoder::container(buf, offset);
```

Comment:
I'll leave it in for now. But, I'm not so convinced that intermediate types is a cleaner api, it can just be confusing to me to have intermediate / helper / conversion types lying around amongst the actual type definitions. Maybe that's just me, but simply encapsulating any conversion logic within the trait seems more clean-cut. But we can leave that discussion for a later pr